### ISSUE TITLE
Lazy.try_force segfaults on multiple domains

### ISSUE THREAD

**Issue by Sudha247** [ocamllabs/ocaml-multicore#483]
This testcase tweaked from `lazy5.ml` in the testsuite, results in a segmentation fault, especially when the number of domains is high. The failure was first observed in this [CI failure](https://github.com/ocaml-multicore/ocaml-multicore/runs/2005271740). The code below fails more reliably than `lazy5.ml`

```ocaml
let num_domains = 16

let rec safe_force l =
  match Lazy.try_force l with
  | Some x -> x
  | None -> (
      Domain.Sync.cpu_relax () ;
      safe_force l
    )

let f count =
  let _n = (Domain.self ():> int) in
  let r = ref 0 in
  for _ = 1 to count do
    incr r;
  done;
  !r

let l = lazy (f 1_000_000_000)
let d1 = Array.init (num_domains - 1) (fun _->
  Domain.spawn (fun () ->
      let _n = (Domain.self ():> int) in
      safe_force l))
let n2 = safe_force l
let n1 = Array.map Domain.join d1

let _ = Printf.printf "n1=%d n2=%d\n" n1.(0) n2
```

**Comment by Sudha247** [comment]
After digging further into the bug, this is what I found. The problematic case is this:

* Thread 1 claims the lazy value and updates its tag from `Lazy_tag` to `Forcing_tag`. After the computation is over, thread 1 updates its tag from `Forcing_tag` to `Forward_tag`.

* Now, some other thread, thread 2 tries to update the lazy value's tag from `Lazy_tag` to `Forcing_tag` (after thread 1 has already updated the tag to `Forward_tag`). This update must have been triggered before thread 1 updated the tag to `Forcing_tag`.

* When thread 2 is trying to update the tag, the `Obj.repr` value of the lazy block returns an incorrect value, leading to a segfault.

So, after the lazy value is forced by one of the threads and the tag updated to `Forward_tag`, the lazy block itself changes somewhere, messing the value in runtime. It's possible that the GC does this, though I didn't find any evidence of that happening whilst stepping through the GC (after (1) and before (2)).

**Comment by kayceesrk** [comment]
As I feared, this is due to the interaction between the implementation of lazy and safe points. :-) Because the lazy values can get short-circuited, the implementation of lazy values assumes that GC does not occur at arbitrary points. The safe points patch only introduces polls at known points. This code has been implemented without consideration for safe points. 

## The problem

Consider these lines

```ocaml
if t = Obj.forward_tag then Some (Obj.obj (Obj.field x 0) : 'arg) else
if t <> Obj.lazy_tag && t <> Obj.forcing_tag then Some (Obj.obj x : 'arg)
else try_force_gen_lazy_block ~only_val lzv
```
from https://github.com/ocaml-multicore/ocaml-multicore/blob/parallel_minor_gc/stdlib/camlinternalLazy.ml#L105-L107. 

The lazy computation can either be not forced earlier (`lazy_tag`) or some thread (including the caller itself) is actively forcing it `lazy_tag` or the lazy computation is completed (`forward_tag`). If the tag is `forward_tag` then the first field of the lazy value contains the result. The GC may also short circuit the reference to the lazy such that reference to the lazy object with the `forward_tag` is replaced by the result (first field of the lazy object). This is why we have the second condition, which casts the object to the target type.

Otherwise, we call `try_force_gen_lazy_block`, which in turn calls `update_tag_forcing`. The comment (precondition) says that we assume that `blk` is a `lazy_tag` object. This is incorrect -- the object is expected to have a `lazy_tag` or `forcing_tag`, but the code is correct **assuming no poll points!**. 

What happens in the buggy case is that the poll point gets triggered at the entry to `update_tag_forcing` and the object gets short-circuited to an integer in this example. Trying to get the `Hd_val` integer value segfaults. 

## Fix

I guess the fix is to repeat the tests
```
if t = Obj.forward_tag then Some (Obj.obj (Obj.field x 0) : 'arg) else
if t <> Obj.lazy_tag && t <> Obj.forcing_tag then Some (Obj.obj x : 'arg)
```
in `update_tag_forcing` which may now additionally return the value as a result. 

## Wider implications

@sadiqj we might take this as an example of how to make code that assumes GC not to run at certain places to be ported to make the code safe under poll points. It would be nice to have a cookbook of recipes that I can follow to update my code such as this one we had for the (no longer needed) C API changes: https://github.com/ocaml-multicore/ocaml-multicore/wiki/C-API-changes. A good place to have that would be in the manual which documents the practices for writing allocation free code, where poll points are inserted, and what one can do to mitigate the issue.

**Comment by sadiqj** [comment]
This is a good spot @kayceesrk.

I think you're right that we need something written up, as with the C API changes. It may be we actually need something for trunk anyway.

(Under the new implementation in the PR against trunk, if you added an always inline to `try_force_gen_lazy_block`, you would avoid this problem)

**Comment by kayceesrk** [comment]
See https://github.com/ocaml-multicore/ocaml-multicore/pull/492.

**Comment by kayceesrk** [comment]
Fixed by #501.
